cloudsync_onedrive.py

```diff
@@ -433,7 +452,7 @@ def _fetch_personal_drives(self):
 
     def _fetch_shared_drives(self):
         # drive items from other drives shared with current user
-        shared = self._direct_api("get", "/me/drive/sharedWithMe")
+        shared = self._safe_direct_api("/me/drive/sharedWithMe", default={})
```

Comment:
does _fetch_personal_drives need this too?

Reply:
I decided against doing this for personal drive fetches because they provide crucial information -- the personal drive id is returned as the unique connection id, and the personal drive is used as the default namespace if a namespace is not explicitly specified by the caller before the connect() call.

So, we can reasonably recover when the site fetch or the shared folder fetch returns garbage, in those cases we can at least fall back on using the personal drive. If the personal drive fetch fails there isn't much we can do -- that is likely a MS account setup issue that the user has to fix.

Reply:
This would have previously raised an exception to the client, which maybe is not ideal, but it would have also prevented the flag which finalizes the various caches from getting set. Now, we will have no shared drives, and no way to detect the problem, and no direct way to repopulate the cache.

Reply:
This change can affect `_get_validated_namespace`, which depends on the `__drive_by_id` cache, producing various `CloudNameSpaceError` exceptions instead of whatever error got swallowed in the `_safe_direct_api`, masking the real cause of the namespace error, especially in the case of `CloudTemporaryError`, but also for every other kind of error too

Reply:
I thought that if the user is trying to connect to the 1 site that failed, it won't show up in the list of namespaces, so rather than raising an exception, it will simply be unavailable in the list of choices with no indication of that it is missing or why.

Perhaps there is a way to provide some indication that an issue arose during the acquisition of the data that failed, so that there could be some UI element that would say, "A OneDrive error occurred while getting the namespaces, if you don't see your namespace here, please retry" or something like that? Simply swallowing the error seems expedient, but if our software returns incomplete results with any frequency, it looks like the problem is with our software, rather than an OneDrive problem

Reply:
Very similar to what I mentioned here: https://github.com/AtakamaLLC/cloudsync-onedrive/pull/38#discussion_r583882005

If you have a namespace ID and you set it before the connect() call and the fetch fails for that particular namespace, you get a namespace error.  But if you are browsing namespaces (list_ns) then there is no namespace error, and items that fail during fetch are missing from the list.

We could collect/store such non-critical errors and expose them via a property to the UI layer or any other interested party.  Does that sound reasonable?

Reply:
Update: now collecting namespace errors in a special namespace that can be queried via `list_ns` (see test)